docs/proposals/secure-impersonation.md

```diff
@@ -0,0 +1,636 @@
+
+Flux APIs must structurally prevent privilege escalation and arbitrary code execution.
+Flux APIs should be congruent with Kubernetes core API's security model.
+
+The implementation should be usable and allow users quick, concise decisions regarding their security needs.
+
+
+## Design
+
+Currently kustomize-controller and helm-controller apply Kustomizations and HelmReleases using their own ServiceAccount by default.
+It's possible to drop privileges using a `serviceAccountName` from the same Namespace -- this is more secure but this is an optional, explicit behavior.
+
+The central controllers should change so they always assume a different User when applying and maintaining objects from manifests in Sources.
+They should never use a controller SA when managing these resulting Objects, even for Garbage Collection, Health-Checks, or "values-from" merging.
+This will allow the audit log to be more granularly examined for controllers' mutations on Flux APIs vs. GitOps-managed objects.
+This technique prevents cross-tenant information disclosure and privilege escalation through misused Health-Checks or Garbage Collection.
+
+#### controller serviceAccounts
+The controller ServiceAccounts are far overprivileged for Flux API operations.
+They should have a ClusterRoleBinding (or RoleBinding for single Namespace installs) for necessary Flux API Kinds and other needed objects such as ConfigMaps/Secrets + a separate ClusterRoleBinding for User Impersonation.
+
+#### user/group names
+All usernames impersonated by kustomize-controller and helm-controller for managing resources will be Namespaced relative to the `Kustomization`/`HelmRelease` object.
+This is similar to the mechanism Kubernetes uses for [ServiceAccount User names](https://kubernetes.io/docs/reference/access-authn-authz/rbac/#referring-to-subjects)).
+Kubernetes uses `system:serviceaccount:{{namespace}}:{{metadata.name}}` for ServiceAccounts.
+The Flux username format should be `flux:user:{{namespace}}:{{spec.user}}`.
+Alternatively flux could use the reserved prefix `system:flux:user:<...>` -- this would break the rules, but make it less likely for other systems to accidentally or maliciously use the FluxUser namespace when authenticating with kube-apiserver.
+The default `spec.user` can be "reconciler".
+A user field will be added to the Kustomization and HelmRelease spec.
+It enables a partial override for the namespaced username the controller uses for that particular `Kustomization`/`HelmRelease`.
+Namespace Owners can use the user override to optionally specify flux `Users` with different `Roles`/`RoleBindings` for managing each workload,
+similar to Pods specifying non-default serviceAccounts from the same Namespace.
+- **Con**: RoleBinding against a FluxUser requires specifying the namespace in the username string which can be hard to do in a generic way with kustomize -- would need to document a kustomize example
+
+Flux controllers will also impersonate two Groups.
```

Comment:
I'm a bit unclear on who is adding the ClusterRoleBinding on the `flux:users` and `flux:users:{{namespace}}` groups. Is the onus on the end-user to create this ClusterRoleBinding after the controllers are deployed or is there some proposed automated process to create these bindings?

Reply:
Please see https://github.com/fluxcd/flux2-multi-tenancy#onboard-tenants

TLDR: there is a `flux create tenant` that creates the user or serviceaccount to be impersonated. At install time, the `flux bootstrap` command will setup the service account and RBAC for the controller pods to enable impersonation.